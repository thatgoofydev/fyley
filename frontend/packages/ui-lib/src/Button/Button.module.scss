@import "src/base-styling/colors";
@import "src/base-styling/vars";

@mixin buttonColor($btn-color) {
  border: 2px solid $btn-color;
  background-color: $btn-color;

  &:hover,&.submitting {
    $hover-color: desaturate(darken($btn-color, 12%), 25%);
    background-color: $hover-color;
    border-color: $hover-color;
  }

  &.secondary {
    color: $btn-color;
  }
}

.button {
  outline: 0;

  padding: $button-normal-padding;
  text-transform: uppercase;
  font-weight: bold;
  letter-spacing: 1px;
  color: $white;
  line-height: $button-line-height;
  border-radius: $button-border-radius;
  min-width: 100px;
  cursor: pointer;
  position: relative;
  transition: background-color .4s, border-color .4s;

  @include buttonColor($blue);

  &.red {
    @include buttonColor($red);
  }

  &:focus {
    border-color: $black;
  }

  &:disabled:not(.submitting):not(.success):not(.error) {
    cursor: default;
    background-color: $grey;
    color: $mid-grey;
    border-color: $grey;
  }

  &.small {
    padding: $button-small-padding;
  }

  &.secondary {
    background-color: transparent;
  }

  &.submitting, &.success, &.error {
    color: transparent;
    cursor: default;

    svg {
      position: absolute;
      top: 50%;
      left: 50%;
      transform: translateX(-50%) translateY(-50%);
      fill: white;
      animation: pulse 0.3s linear 1;
    }
  }

  &.submitting::before {
    content: "";
    position: absolute;
    top: 50%;
    left: 50%;
    width: 20px;
    height: 20px;
    transform: translateX(-50%) translateY(-50%);

    border: 3px solid rgba(255, 255, 255, .35);
    border-right: 3px solid white;
    border-radius: 50%;
    animation: rotate .8s linear infinite;
  }

  &.small.submitting::before {
    width: 16px;
    height: 16px;
    border-width: 2px;
  }

  &.success {
    background-color: $green;
    border-color: $green;
  }

  &.error {
    background-color: $red;
    border-color: $red;
  }
}

@keyframes rotate {
  0% {
    transform: translateX(-50%) translateY(-50%) rotate(0deg);
  }
  100% {
    transform: translateX(-50%) translateY(-50%) rotate(360deg);
  }
}

@keyframes pulse {
  25%{
    transform: translateX(-50%) translateY(-50%) scale(0.8);
  }
  75% {
    transform: translateX(-50%) translateY(-50%) scale(1.3);
  }
}