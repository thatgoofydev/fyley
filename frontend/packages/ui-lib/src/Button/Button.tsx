import React, { FunctionComponent } from "react";
import styles from "./Button.module.scss";
import classNames from "classnames";

export interface IButtonProps {
  label: string;
  style?: "primary" | "secondary";
  type?: "submit" | "button" | "reset";
  size?: "normal" | "small";
  color?: "blue" | "red";
  disabled?: boolean;
  actionState?: "submitting" | "success" | "error" | "none";
}

export const Button: FunctionComponent<IButtonProps> = ({
  label,
  style = "primary",
  type = "button",
  size = "normal",
  color = "blue",
  disabled,
  actionState = "none"
}) => {
  const classes: string = classNames(
    styles.button,
    style === "secondary" ? styles.secondary : undefined,
    size === "small" ? styles.small : undefined,
    styles[color],
    actionState ? styles[actionState] : undefined
  );

  const svgSize = size === "small" ? 18 : 24;
  return (
    <button type={type} disabled={disabled} className={classes}>
      {label}
      {actionState === "success" && (
        <svg
          xmlns="http://www.w3.org/2000/svg"
          width={svgSize}
          height={svgSize}
          viewBox="0 0 24 24"
        >
          <path d="M20.3 2L9 13.6l-5.3-5L0 12.3 9 21 24 5.7z" />
        </svg>
      )}
      {actionState === "error" && (
        <svg
          xmlns="http://www.w3.org/2000/svg"
          width={svgSize}
          height={svgSize}
          viewBox="0 0 24 24"
        >
          <path d="M24 20.2L15.7 12l8.2-8.3L20.2 0 12 8.3 3.7.1 0 3.8 8.3 12 .1 20.3 3.8 24l8.2-8.3 8.3 8.2z" />
        </svg>
      )}
    </button>
  );
};
