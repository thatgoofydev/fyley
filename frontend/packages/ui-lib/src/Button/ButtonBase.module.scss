@import "src/base-styling/colors";

$button-colors: ("danger": $red, "warning": $orange);
$bg-color: #FAFBFC;

.button {
  border-radius: 4px;
  background-color: $bg-color;
  border: 1px solid darken($bg-color, 15%);
  padding: 2px 14px;
  line-height: 20px;
  color: $primary-text;
  cursor: pointer;

  &:hover {
    background-color: darken($bg-color, 5%);
  }

  &:disabled {
    background-color: rgba($bg-color, 0.6);
    color: rgba($primary-text, 0.4);
    cursor: not-allowed;
  }

  &:active {
    background-color: lighten($bg-color, 10%);
  }

  &:focus {
    outline: 0;
    border: 1px solid darken($bg-color, 40%);
    box-shadow: 0 0 2px darken($bg-color, 50%);
  }

  @each $name, $color in $button-colors {
    &.#{$name} {
      color: $color;

      &:hover {
        color: white;
        background-color: $color;
      }

      &:active {
        color: white;
        background-color: lighten($color, 10%);
      }
    }
  }
}