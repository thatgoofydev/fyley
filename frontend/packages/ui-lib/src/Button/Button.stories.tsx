import React from "react";
import { Meta, Story } from "@storybook/react";
import { Button, IButtonProps } from "./Button";

export default {
  title: "Buttons/Button",
  component: Button
} as Meta;

export const ButtonStory: Story<IButtonProps> = args => {
  return <Button {...args} />;
};
ButtonStory.storyName = "Button";
ButtonStory.args = {
  label: "Button",
  style: "primary",
  size: "normal",
  type: "button",
  color: "blue",
  disabled: false
};
