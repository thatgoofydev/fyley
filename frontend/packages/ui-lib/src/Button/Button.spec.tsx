import React from "react";
import { render, cleanup } from "@testing-library/react";
import { Button, IButtonProps } from "./Button";

const noop = () => {};
const defaultProps: IButtonProps = {
  label: "Button",
  style: "primary",
  type: "button",
  size: "normal",
  color: "blue",
  disabled: false,
  onClick: noop
};

afterEach(cleanup);

describe("Button", function () {
  describe("should render correctly", function () {
    it("when default", function () {
      const { container } = render(<Button {...defaultProps} />);
      expect(container).toMatchSnapshot();
    });

    it("when small size", function () {
      const props: IButtonProps = {
        ...defaultProps,
        size: "small"
      };
      const { container } = render(<Button {...props} />);
      expect(container).toMatchSnapshot();
    });

    it("when disabled", function () {
      const props: IButtonProps = {
        ...defaultProps,
        disabled: true
      };
      const { container } = render(<Button {...props} />);
      expect(container).toMatchSnapshot();
    });

    it("when secondary style", function () {
      const props: IButtonProps = {
        ...defaultProps,
        style: "secondary"
      };
      const { container } = render(<Button {...props} />);
      expect(container).toMatchSnapshot();
    });

    // TODO add tests with form to check submit feedback
  });

  it("should trigger onClick", function () {
    const onClickCallback = jest.fn();

    const props: IButtonProps = {
      ...defaultProps,
      onClick: onClickCallback,
      "data-testid": "btn"
    };

    const { getByTestId } = render(<Button {...props} />);
    getByTestId("btn").click();

    expect(onClickCallback).toHaveBeenCalledTimes(1);
  });
});
