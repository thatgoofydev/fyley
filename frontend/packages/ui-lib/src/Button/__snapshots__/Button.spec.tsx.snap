// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render correctly when default 1`] = `
<div>
  <button
    class="button blue"
    type="button"
  >
    Button
  </button>
</div>
`;

exports[`Button should render correctly when disabled 1`] = `
<div>
  <button
    class="button blue"
    disabled=""
    type="button"
  >
    Button
  </button>
</div>
`;

exports[`Button should render correctly when secondary style 1`] = `
<div>
  <button
    class="button secondary blue"
    type="button"
  >
    Button
  </button>
</div>
`;

exports[`Button should render correctly when small size 1`] = `
<div>
  <button
    class="button small blue"
    type="button"
  >
    Button
  </button>
</div>
`;
